File: csharp/784.letter-case-permutation.cs
/*
 * @lc app=leetcode id=784 lang=csharp
 *
 * [784] Letter Case Permutation
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    private static void LetterCasePermutationRecurse(List<string> ret, char[] str, int index) {
        if (index == str.Length) {
            ret.Add(new string(str));
            return;
        }
        LetterCasePermutationRecurse(ret, str, index + 1);
        char c = str[index];
        if (c >= 'a' && c <= 'z') {
            str[index] = (char) ((int) c + (int) 'A' - (int) 'a');
            LetterCasePermutationRecurse(ret, str, index + 1);
            str[index] = c;
        } else if (c >= 'A' && c <= 'Z') {
            str[index] = (char) ((int) c - (int) 'A' + (int) 'a');
            LetterCasePermutationRecurse(ret, str, index + 1);
            str[index] = c;
        }
    }
    
    public IList<string> LetterCasePermutation(string S) {
        var ret = new List<string>();
        LetterCasePermutationRecurse(ret, S.ToCharArray(), 0);
        return ret;
    }
}
// @lc code=end

File: java/784.letter-case-permutation.java
/*
 * @lc app=leetcode id=784 lang=java
 *
 * [784] Letter Case Permutation
 */

// @lc code=start
class Solution {
    
    private static void letterCasePermutation(String S, int i, List<String> ret, StringBuilder sb) {
        if (i == S.length()) {
            ret.add(sb.toString());
            return;
        }
        char c = S.charAt(i);
        sb.append(c);
        letterCasePermutation(S, i + 1, ret, sb);
        sb.deleteCharAt(i);
        if (c >= 'a' && c <= 'z') {
            sb.append((char)((int)c - (int)'a' + (int)'A'));
            letterCasePermutation(S, i + 1, ret, sb);
            sb.deleteCharAt(i);
        } else if (c >= 'A' && c <= 'Z') {
            sb.append((char)((int)c - (int)'A' + (int)'a'));
            letterCasePermutation(S, i + 1, ret, sb);
            sb.deleteCharAt(i);
        }
    }

    public List<String> letterCasePermutation(String S) {
        List<String> ret = new ArrayList<String>();
        letterCasePermutation(S, 0, ret, new StringBuilder(S.length()));
        return ret;
    }
}
// @lc code=end

